lib/windows/app/actions/deviceActions.js

```diff
@@ -0,0 +1,313 @@
+ */
+export function startWatchingDevices() {
+    const config = getAppConfig();
+
+    return (dispatch, getState) => {
+        deviceLister = new DeviceLister(config.selectorTraits);
+
+        deviceLister.on('conflated', devices => {
+            const state = getState();
+            if (state.core.device.selectedSerialNumber !== null &&
+                !devices.has(state.core.device.selectedSerialNumber)) {
+                dispatch(deselectDevice());
+            }
+
+            dispatch(devicesDetectedAction(Array.from(devices.values())));
+        });
+        deviceLister.on('error', logDeviceListerError);
+        deviceLister.start();
+    };
+}
+
+/**
+ * Stops watching for devices.
+ *
+ * @returns {function(*)} Function that can be passed to redux dispatch.
+ */
+export function stopWatchingDevices() {
+    return () => {
+        deviceLister.removeAllListeners('conflated');
+        deviceLister.removeAllListeners('error');
+        deviceLister.stop();
+    };
+}
+
+/**
+ * Asks the user to provide input during device setup. If a list of choices are
+ * given, and the user selects one of them, then then promise will resolve with
+ * the selected value. If no choices are given, and the user confirms, then the
+ * promise will just resolve with true. Will reject if the user cancels.
+ *
+ * @param {function} dispatch The redux dispatch function.
+ * @param {String} message The message to display to the user.
+ * @param {Array<String>} [choices] The choices to display to the user (optional).
+ * @returns {Promise<String>} Promise that resolves with the user input.
+ */
+function getDeviceSetupUserInput(dispatch, message, choices = undefined) {
+    return new Promise((resolve, reject) => {
+        deviceSetupCallback = choice => {
+            if (choice) {
+                resolve(choice);
+            } else {
+                reject(new Error('Cancelled by user.'));
+            }
+        };
+        dispatch(deviceSetupInputRequiredAction(message, choices));
+    });
+}
+
+/**
+ * Selects a device and sets it up for use according to the `config.deviceSetup`
+ * configuration given by the app.
+ *
+ * @param {Object} device Device object, ref. nrf-device-lister.
+ * @returns {function(*)} Function that can be passed to redux dispatch.
+ */
+export function selectAndSetupDevice(device) {
+    return dispatch => {
+        dispatch(deviceSelectedAction(device));
+
+        const config = getAppConfig();
+        if (config.deviceSetup) {
+            // During device setup, the device may go in and out of bootloader
+            // mode. This will make it appear as detached in the device lister,
+            // causing a DESELECT_DEVICE. To avoid this, we stop the device
+            // lister while setting up the device, and start it again after the
+            // device has been set up.
+            dispatch(stopWatchingDevices());
+
+            const deviceSetup = { ...config.deviceSetup };
```

Comment:
Why not just `const deviceSetup = config.deviceSetup;` ? 
:thinking: 
Is there a need to create a shallow copy somehow?